File: J9Ch/src/J_4_Breadth_First_Search/Optional_6/python_cpp/_624_Remove_Substrings_Medium.py
'''/**
 * 624 Remove Substrings'''


class Solution:
    # @param {string} s a string
    # @param {set} dict a set of n substrings
    # @return {int} the minimum length
    def minLength(self, s, dict):
        # Write your code here
        import Queue
        que = Queue.Queue()
        que.put(s)
        hash = set([s])

        min = len(s)

        while not que.empty():
            s = que.get()
            for sub in dict:
                found = s.find(sub)
                while found != -1:
                    new_s = s[:found] + s[found + len(sub):]
                    if new_s not in hash:
                        if len(new_s) < min:
                            min = len(new_s)
                        que.put(new_s)
                        hash.add(new_s)

                    found = s.find(sub, found + 1)
        return min

File: J9Ch/src/J_4_Breadth_First_Search/Optional_6/python_cpp/_624_Remove_Substrings_Medium.cpp
/**
 * 624 Remove Substrings*/


class Solution {
public:
    /**
     * @param s a string
     * @param dict a set of n substrings
     * @return the minimum length
     */
    int minLength(string& s, set<string>& dict) {
        // Write your code here
        queue<string> que;
        unordered_set<string> hash;

        int min = s.size();
        que.push(s);
        hash.insert(s);

        while (!que.empty()) {
            string s = que.front();
            que.pop();
            for (auto& sub : dict) {
                int found = s.find(sub);
                while (found != -1) {
                    string new_s = s.substr(0, found) +
                        s.substr(found + sub.size(), s.size() - found - sub.size());
                    if (hash.find(new_s) == hash.end()) {
                        if (new_s.size() < min)
                            min = new_s.size();
                        que.push(new_s);
                        hash.insert(new_s);
                    }
                    found = s.find(sub, found + 1);
                }
            }
        }
        return min;
    }
};